cmd/umoci/raw-unpack.go

```diff
@@ -0,0 +1,171 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/apex/log"
+	"github.com/openSUSE/umoci/oci/cas/dir"
+	"github.com/openSUSE/umoci/oci/casext"
+	"github.com/openSUSE/umoci/oci/layer"
+	"github.com/openSUSE/umoci/pkg/idtools"
+	ispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/net/context"
+)
+
+var rawUnpackCommand = cli.Command{
+	Name:  "unpack",
+	Usage: "unpacks a reference into a rootfs",
+	ArgsUsage: `--image <image-path>[:<tag>] <rootfs>
+
+Where "<image-path>" is the path to the OCI image, "<tag>" is the name of the
+tagged image to unpack (if not specified, defaults to "latest") and "<rootfs>"
+is the destination to unpack the image to.`,
+
+	// unpack reads manifest information.
+	Category: "image",
+
+	Flags: []cli.Flag{
+		cli.StringSliceFlag{
+			Name:  "uid-map",
+			Usage: "specifies a uid mapping to use when repacking (container:host:size)",
+		},
+		cli.StringSliceFlag{
+			Name:  "gid-map",
+			Usage: "specifies a gid mapping to use when repacking (container:host:size)",
+		},
+		cli.BoolFlag{
+			Name:  "rootless",
+			Usage: "enable rootless unpacking support",
+		},
+	},
+
+	Action: rawUnpack,
+
+	Before: func(ctx *cli.Context) error {
+		if ctx.NArg() != 1 {
+			return errors.Errorf("invalid number of positional arguments: expected <rootfs>")
+		}
+		if ctx.Args().First() == "" {
+			return errors.Errorf("rootfs path cannot be empty")
+		}
+		ctx.App.Metadata["rootfs"] = ctx.Args().First()
+		return nil
+	},
+}
+
+func rawUnpack(ctx *cli.Context) error {
+	imagePath := ctx.App.Metadata["--image-path"].(string)
+	fromName := ctx.App.Metadata["--image-tag"].(string)
+	rootfsPath := ctx.App.Metadata["rootfs"].(string)
+
+	var meta UmociMeta
+	meta.Version = UmociMetaVersion
+
+	// Parse map options.
+	// We need to set mappings if we're in rootless mode.
+	meta.MapOptions.Rootless = ctx.Bool("rootless")
+	if meta.MapOptions.Rootless {
+		if !ctx.IsSet("uid-map") {
+			ctx.Set("uid-map", fmt.Sprintf("0:%d:1", os.Geteuid()))
+		}
+		if !ctx.IsSet("gid-map") {
+			ctx.Set("gid-map", fmt.Sprintf("0:%d:1", os.Getegid()))
+		}
+	}
+	// Parse and set up the mapping options.
+	for _, uidmap := range ctx.StringSlice("uid-map") {
+		idMap, err := idtools.ParseMapping(uidmap)
+		if err != nil {
+			return errors.Wrapf(err, "failure parsing --uid-map %s: %s", uidmap)
+		}
+		meta.MapOptions.UIDMappings = append(meta.MapOptions.UIDMappings, idMap)
+	}
+	for _, gidmap := range ctx.StringSlice("gid-map") {
+		idMap, err := idtools.ParseMapping(gidmap)
+		if err != nil {
+			return errors.Wrapf(err, "failure parsing --gid-map %s: %s", gidmap)
+		}
+		meta.MapOptions.GIDMappings = append(meta.MapOptions.GIDMappings, idMap)
+	}
```

Comment:
Since this code is now used by `raw unpack` (and will be used by `insert` in #237) this should be moved to a separate function.